aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,148 @@
+    public entry fun add_nonce_bucket(aptos_framework: &signer) acquires NonceHistory {
+        if (exists<NonceHistory>(@aptos_framework)) {
+            let current_time = timestamp::now_seconds();
+            let nonce_history = borrow_global_mut<NonceHistory>(@aptos_framework);
+            if (!table::contains(&nonce_history.nonce_table, nonce_history.next_key)) {
+                let nonces = vector::empty();
+                let last_stored_times = vector::empty();
+                vector::push_back(&mut nonces, big_ordered_map::new());
+                vector::push_back(&mut nonces, big_ordered_map::new());
+                vector::push_back(&mut last_stored_times, 0);
+                vector::push_back(&mut last_stored_times, 0);
+                table::add(&mut nonce_history.nonce_table, nonce_history.next_key, Bucket {
+                    nonces: nonces,
+                    last_stored_times: last_stored_times,
+                });
+            };
+            nonce_history.next_key = nonce_history.next_key + 1;
+        };
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        let nonce_history = borrow_global_mut<NonceHistory>(@aptos_framework);
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let map_index = (txn_expiration_time / MAP_SWITCH_TIME) % 2;
+        let current_time = timestamp::now_seconds();
+        if (!table::contains(&nonce_history.nonce_table, index)) {
+            let nonces = vector::empty();
+            let last_stored_times = vector::empty();
+            vector::push_back(&mut nonces, big_ordered_map::new());
+            vector::push_back(&mut nonces, big_ordered_map::new());
+            vector::push_back(&mut last_stored_times, 0);
+            vector::push_back(&mut last_stored_times, 0);
+            big_ordered_map::add(&mut nonces[map_index], nonce_key, txn_expiration_time);
+            last_stored_times[map_index] = current_time;
+            table::add(&mut nonce_history.nonce_table, index, Bucket {
+                nonces: nonces,
+                last_stored_times: last_stored_times,
+            });
+            return true;
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, index);
+        // Question[Orderless]: As this is an expensive operation, didn't do it for both buckets.
+        // Should we do this for both buckets? 
+        if (bucket.last_stored_times[1-map_index] + MAX_EXPIRATION_TIME < current_time) {
```

Comment:
this code is not very readable, looks it's switching the 0, 1 tables with some gc logics but it's mixing with existence checks as well